cli/src/semgrep/lsp/convert.py

```diff
@@ -14,47 +25,107 @@ def findings_severity_to_diagnostic(severity: RuleSeverity) -> DiagnosticSeverit
+            get_metavar_related(m, MetavarValue.from_json(d))
+            for m, d in rule_match.extra["metavars"].items()
         )
+    return []
 
-    rule_url = f"{env.semgrep_url}/r/{finding['check_id']}"
 
-    diagnostic = {
+# Convert a rule match to a LSP diagnostic
+# According to https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic
+def rule_match_to_diagnostic(rule_match: RuleMatch) -> JsonObject:
+    env = get_state().env
+    rule_url = f"{env.semgrep_url}/r/{rule_match.rule_id}"
+    diagnostic: MutableMapping[str, Any] = {
         "range": {
-            "start": {"line": start_line, "character": start_col},
+            # We start at 0, but the LSP starts at 1
+            "start": {
+                "line": rule_match.start.line - 1,
+                "character": rule_match.start.col - 1,
+            },
             "end": {
-                "line": end_line,
-                "character": end_col,
+                "line": rule_match.end.line - 1,
+                "character": rule_match.end.col - 1,
             },
         },
-        "message": finding["extra"]["message"],
-        "severity": findings_severity_to_diagnostic(
-            RuleSeverity(finding["extra"]["severity"])
-        ).value,
+        "message": rule_match.message,
+        "severity": findings_severity_to_diagnostic(rule_match.severity).value,
         "source": "Semgrep",
-        "code": finding["check_id"],
+        "code": rule_match.rule_id,
         "codeDescription": {"href": rule_url},
-        "data": {"matchSource": match_source},
+        "data": {
+            "matchSource": rule_match.syntactic_context,
+            "uri": f"file://{rule_match.path}",
+        },
+        "relatedInformation": rule_match_get_related(rule_match),
     }
 
-    if "fix" in finding["extra"]:
-        diagnostic["data"]["fix"] = finding["extra"]["fix"]
-    if "fix_regex" in finding["extra"]:
-        diagnostic["data"]["fix_regex"] = finding["extra"]["fix_regex"]
+    fix_message = None
+    if rule_match.extra.get("metavars") is not None:
+        diagnostic["data"]["metavars"] = rule_match.extra["metavars"]
+    if rule_match.fix:
+        diagnostic["data"]["fix"] = rule_match.fix
+        fix_message = rule_match.fix
+    if rule_match.fix_regex:
+        diagnostic["data"]["fix_regex"] = rule_match.fix_regex
+        fix_message = rule_match.fix
 
+    if fix_message is not None:
+        diagnostic["message"] += f"\nFix: {fix_message}"
+    # This looks better
+    diagnostic["message"] += "\n"
     return diagnostic
+
+
+def rule_match_map_to_diagnostics(rule_map: RuleMatchMap) -> List[JsonObject]:
+    return flatten(
+        [
+            list(map(rule_match_to_diagnostic, rule_match_list))
+            for rule_match_list in rule_map.values()
+        ]
+    )
+
+
+# get assocaitated files for a rule match
+def rule_to_files(rule: Rule, target_manager: TargetManager) -> List[str]:
```

Comment:
```suggestion
def rule_to_files(rule: Rule, target_manager: TargetManager) -> List[str]:
    """Get associated files for a rule match."""
```

^ there are a lot of docs automation tools that expect functions to be documented in this format. These docstrings are also made available when e.g. hovering over a call to `rule_to_files` in vscode, or as `rule_to_files.__doc__` in an interpreter

Reply:
please update the other functions as well according to this if you have time!